File: 프로그래머스/2/340212. ［PCCP 기출문제］ 2번 ／ 퍼즐 게임 챌린지/［PCCP 기출문제］ 2번 ／ 퍼즐 게임 챌린지.java
class Solution {
    public long solution(int[] diffs, int[] times, long limit) {
        int size = diffs.length;
        long left = 1;
        long right = 1_000_000_000_000_000L;
        long mid = -1;
        
        
        while(left <= right){
            mid = (left + right) / 2;
            long timeSpent = 0;
            
            for(int i = 0; i < size; i++) {
                if(diffs[i] <= mid) {
                    timeSpent += times[i];
                } else {
                    timeSpent += (diffs[i] - mid) * (times[i] + times[i-1]) + times[i];
                }
            }
            
            System.out.println(mid+" "+timeSpent);
            
            if(timeSpent <= limit) {
                right = mid - 1;
            } else {
                left = mid + 1;
            }
        }
        return left;
    }
}

File: 프로그래머스/2/340212. ［PCCP 기출문제］ 2번 ／ 퍼즐 게임 챌린지/［PCCP 기출문제］ 2번 ／ 퍼즐 게임 챌린지.py
import sys
input = sys.stdin.readline
def solution(diffs, times, limit):
    diffs.insert(0,0)
    times.insert(0,0)
    puzzleCnt = len(diffs)
    levelEnd = max(diffs)
    start = 1
    end = levelEnd
    level = -1
    result = -1
    answer = []

    while start <= end :
        level = (start + end) // 2
        time = 0
        for j in range(1,puzzleCnt) :
            if diffs[j] <= level :
                time += times[j]
            else :
                time += (((diffs[j] - level) * (times[j] + times[j-1])) + times[j])

        if time <= limit :
            end = level - 1
            result = level
        else :
            start = level + 1

    return result